cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -16,10 +16,235 @@
+                     md5_shift_constants_type const* shift_constants)
+    : d_hash_constants(hash_constants), d_shift_constants(shift_constants)
+  {
+  }
+
+  /**
+   * @brief Core MD5 algorithm implementation. Processes a single 512-bit chunk,
+   * updating the hash value so far. Does not zero out the buffer contents.
+   */
+  void __device__ hash_step(md5_intermediate_data* hash_state) const
+  {
+    uint32_t A = hash_state->hash_value[0];
+    uint32_t B = hash_state->hash_value[1];
+    uint32_t C = hash_state->hash_value[2];
+    uint32_t D = hash_state->hash_value[3];
+
+    for (unsigned int j = 0; j < 64; j++) {
+      uint32_t F;
+      uint32_t g;
+      switch (j / 16) {
+        case 0:
+          F = (B & C) | ((~B) & D);
+          g = j;
+          break;
+        case 1:
+          F = (D & B) | ((~D) & C);
+          g = (5 * j + 1) % 16;
+          break;
+        case 2:
+          F = B ^ C ^ D;
+          g = (3 * j + 5) % 16;
+          break;
+        case 3:
+          F = C ^ (B | (~D));
+          g = (7 * j) % 16;
+          break;
+      }
+
+      uint32_t buffer_element_as_int;
+      std::memcpy(&buffer_element_as_int, hash_state->buffer + g * 4, 4);
+      F = F + A + d_hash_constants[j] + buffer_element_as_int;
+      A = D;
+      D = C;
+      C = B;
+      B = B + __funnelshift_l(F, F, d_shift_constants[((j / 16) * 4) + (j % 4)]);
+    }
+
+    hash_state->hash_value[0] += A;
+    hash_state->hash_value[1] += B;
+    hash_state->hash_value[2] += C;
+    hash_state->hash_value[3] += D;
+
+    hash_state->buffer_length = 0;
+  }
+  /**
+   * @brief Core MD5 element processing function
+   */
+  template <typename TKey>
+  void __device__ process(TKey const& key, md5_intermediate_data* hash_state) const
+  {
+    uint32_t const len  = sizeof(TKey);
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < 64) {
+      thrust::copy_n(thrust::seq, data, len, hash_state->buffer + hash_state->buffer_length);
+      hash_state->buffer_length += len;
+    } else {
+      uint32_t copylen = 64 - hash_state->buffer_length;
+
+      thrust::copy_n(thrust::seq, data, copylen, hash_state->buffer + hash_state->buffer_length);
+      hash_step(hash_state);
+
+      while (len > 64 + copylen) {
+        thrust::copy_n(thrust::seq, data + copylen, 64, hash_state->buffer);
+        hash_step(hash_state);
+        copylen += 64;
+      }
+
+      thrust::copy_n(thrust::seq, data + copylen, len - copylen, hash_state->buffer);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  void __device__ finalize(md5_intermediate_data* hash_state, char* result_location) const
+  {
+    auto const full_length = (static_cast<uint64_t>(hash_state->message_length)) << 3;
+    thrust::fill_n(thrust::seq, hash_state->buffer + hash_state->buffer_length, 1, 0x80);
+
+    if (hash_state->buffer_length < 56) {
```

Comment:
My comment about hardcoded numbers has not been resolved. There might be other comments that got hidden when the code moved, will have to check manually.

Reply:
Should be fixed now